### PR TITLE
fix(php8) - promise interface & react update

### LINKED ISSUE #19087: Composer v2.6.1 install fatal error
### Operating System

Ubuntu 20.04.6 LTS

### Programming Languages

PHP

### CCXT Version

4.0.80

### Description

When I run `composer install --verbose --prefer-dist --no-progress --no-interaction --no-dev --optimize-autoloader` on my production server I get a PHP fatal error. I did not have this error before and I've run composer many times on production. Looks like the issue started with version 4.0.75 or 4.0.77. But now even if I lower the version still get the error 

 I don't get this error on my local environment when I run composer install (using docker)

`"require": {
        "ccxt/ccxt": "^4.0.80",
  }`

### Code

```
  PHP Fatal error:  Declaration of ccxt\pro\Future::then($onFulfilled = null,  
   $onRejected = null, $onProgress = null) must be compatible with React\Prom  
  ise\PromiseInterface::then(?callable $onFulfilled = null, ?callable $onReje  
  cted = null): React\Promise\PromiseInterface in /var/www/.../vendor/ccxt/ccxt/php/pro/Future.php on line 15 

```

### LINKED ISSUE #19310: Cannot Update or Install CCXT on php v8.2.10
### Operating System

Ubuntu 22.04

### Programming Languages

PHP

### CCXT Version

4.0.99

### Description

Cannot update or install ccxt on php v8.2.10
Fatal error: Declaration of ccxt\pro\Future::then($onFulfilled = null, $onRejected = null, $onProgress = null) must be compatible with React\Promise\PromiseInterface::then(?callable $onFulfilled = null, ?callable $onRejected = null): React\Promise\PromiseInterface in /var/www/sleep.finance/vendor/ccxt/ccxt/php/pro/Future.php on line 15

### Code

```
  lock file operations: 0 installs, 1 update, 0 removals
  - Upgrading ccxt/ccxt (4.0.98 => 4.0.99)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Downloading ccxt/ccxt (4.0.99)
  - Upgrading ccxt/ccxt (4.0.98 => 4.0.99): Extracting archive
Package fgrosse/phpasn1 is abandoned, you should avoid using it. No replacement was suggested.
Package laravelcollective/html is abandoned, you should avoid using it. Use spatie/laravel-html instead.
Package php-http/message-factory is abandoned, you should avoid using it. Use psr/http-factory instead.
Generating optimized autoload files
Class ccxt\async\Throttler located in ./vendor/ccxt/ccxt/php/async/Throttle.php does not comply with psr-4 autoloading standard. Skipping.
Class ccxt\baseMainTestClass located in ./vendor/ccxt/ccxt/php/test/test_sync.php does not comply with psr-4 autoloading standard. Skipping.
Class ccxt\testMainClass located in ./vendor/ccxt/ccxt/php/test/test_sync.php does not comply with psr-4 autoloading standard. Skipping.
Class ccxt\baseMainTestClass located in ./vendor/ccxt/ccxt/php/test/test_async.php does not comply with psr-4 autoloading standard. Skipping.
Class ccxt\testMainClass located in ./vendor/ccxt/ccxt/php/test/test_async.php does not comply with psr-4 autoloading standard. Skipping.
Class ccxt\async\Throttler located in ./vendor/ccxt/ccxt/php/async/Throttle.php does not comply with psr-4 autoloading standard. Skipping.
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
PHP Fatal error:  Declaration of ccxt\pro\Future::then($onFulfilled = null, $onRejected = null, $onProgress = null) must be compatible with React\Promise\PromiseInterface::then(?callable $onFulfilled = null, ?callable $onRejected = null): React\Promise\PromiseInterface in /var/www/sleep.finance/vendor/ccxt/ccxt/php/pro/Future.php on line 15

Fatal error: Declaration of ccxt\pro\Future::then($onFulfilled = null, $onRejected = null, $onProgress = null) must be compatible with React\Promise\PromiseInterface::then(?callable $onFulfilled = null, ?callable $onRejected = null): React\Promise\PromiseInterface in /var/www/sleep.finance/vendor/ccxt/ccxt/php/pro/Future.php on line 15

```

### PR DESCRIPTION
**Pull request by ttodua**
reactphp-promise is now of v3.0.0 version, where `then()` has signature of: https://github.com/reactphp/promise/blob/3.x/src/PromiseInterface.php#L40

while we had older signature:
https://github.com/ccxt/ccxt/blob/master/php/pro/Future.php#L15

which caused an issue ( fix #19087 fix #19310 ).

### CODE REVIEW

**Comment by carlosmiei** [comment]
@ttodua can you check this error? 
image

**Comment by ttodua** [comment]
> @ttodua can you check this error? image

hm, the latest reactPhp library does not include that anymore. let me check this, where we retain old version reference, might be -lock file

**Comment by ttodua** [comment]
@carlosmiei does something comes to your mind why this might be happening?

**Comment by ttodua** [comment]
still this file https://github.com/reactphp/promise/blob/2.x/src/PromiseInterface.php was being used, but i've enforced v3 version, lmk if thats  acceptable

**Comment by ttodua** [comment]
hm, still does not helped. @carlosmiei lmk if you have any idea

**Comment by sc0Vu** [comment]
@carlosmiei @ttodua 

It looks like composer didn't download new promise package first time:

image

After I use require command `composer require react/promise 3.0.0`, the package was updated to real v3:

image

BTW, I think composer identify v2.1.0 (f913fb8) commit as v3.0.0

**Comment by ttodua** [comment]
@sc0Vu so what do you suggest, do you know quick workaround for this? otherwise, I'll google into the subject

**Comment by miguilimzero** [comment]
The problem here is that composer 2.6 is resolving the `react/promise` wrong version. That's why "sometimes" works as expected and "sometimes" does not. It depends on which composer version you're using. 

This PR aims to fix this: https://github.com/composer/composer/pull/11634. 

If the composer PRs get merged, this problem should be fixed. 

This current PR can be merged without problem (updating ccxt to use react/promise v3) or this PR can be closed (keeping the react/promise v2) as the issue is related to the composer and not to ccxt nor react/promise.

File: composer.json [hunk 1]
```diff
@@ -240,7 +240,7 @@
         "ratchet/pawl": ">=0.4.1",
         "react/http": ">=1.6.0",
         "react/async": "^4.0.0",
-        "react/promise": "^2.9.0",
+        "react/promise": ">=3.0.0",
```
Comment (Seldaek) [inline composer.json:243]: Why not this?

```suggestion
        "react/promise": "^3.0.0",
```

>= is unbounded and dangerous

Comment (ttodua) [reply composer.json]: right, i've had `^` initially, just PR build didn't install v3 and I've just switched to that to test. however, switched back now as you pointed out.

**Comment by ttodua** [comment]
> The problem here is that composer 2.6 is resolving the `react/promise` wrong version. That's why "sometimes" works as expected and "sometimes" does not. It depends on which composer version you're using.
> 
> This PR aims to fix this: [composer/composer#11634](https://github.com/composer/composer/pull/11634).
> 
> If the composer PRs get merged, this problem should be fixed.
> 
> This current PR can be merged without problem (updating ccxt to use react/promise v3) or this PR can be closed (keeping the react/promise v2) as the issue is related to the composer and not to ccxt nor react/promise.

thanks for comment, I agree with you. the only nuance is that from this PR and onwards, 'promise v3' changed signatures of some methods (breaking changes), and ccxt codes are being also updated to comply with changed signatures. so, we can no longer keep v2 versions

**Comment by ttodua** [comment]
@carlosmiei @sc0Vu please post your thoughts on this PR, whether to merge or not

**Comment by sc0Vu** [comment]
@miguilimzero intersting finding, thanks!

@carlosmiei  @ttodua composer team already merged fix and it work for me.

**Comment by miguilimzero** [comment]
It seems the composer PR didn't fix the issue. I updated to Composer v2.6.3 and tried the following steps:

- `composer clear-cache`
- Deleted `vendor` folder
- Deleted `composer.lock` file
- `composer install`

Although my `composer.lock` shows the correct react/promise version with the correct release commit, it still downloads the v3.0 for some reason:

image

Following the distribution zip URL downloads the correct version with the correct files.

**Comment by miguilimzero** [comment]
I will try to create a clean repository to reproduce the bug to create an issue in the composer repository.

**Comment by miguilimzero** [comment]
I just created an issue in the composer repository about this problem: https://github.com/composer/composer/issues/11654

**Comment by ttodua** [comment]
@miguilimzero thanks for the input! I'll wait what they will respond to your issue 👍

**Comment by c0dehunter** [comment]
Is there anything we can co from our side so at least we can install CCXT 4.* on our machines?

**Comment by indigoram89** [comment]
> @sc0Vu thanks for follow-up ! however, I still think this PR should be merged, because as I said in top-post, the signature is changed and we should also update it. @carlosmiei

We need to do It

**Comment by miguilimzero** [comment]
@ttodua So, the composer maintainer explained that this happens because the composer's dependencies conflict with ccxt dependencies when doing a static method call in a script.

Composer recently updated from react/promise v2 to v3, which broke the package here. 

When the composer calls a static method in a script, this call binds the composer dependencies and overwrites the react/promise v2 from the project dependencies with the v3 from the composer itself, throwing the error on the composer install.

Laravel, by default, has a post-autoload-dump script in the composer.json that calls the method  `Illuminate\Foundation\ComposerScripts::postAutoloadDump`.  So, this update broke ccxt package for all Laravel projects.

From what I understood from the maintainer feedback, the only way to fix this and be compatible with Laravel projects from ccxt side is to bump the react/promise to v3.

@c0dehunter A way to get around this issue is to remove the script line from your composer.json. This line clears the configs, services, and packages cached files on every composer install. You can create a command in your project that mimics this script with the following code:

```php
$laravel = new \Illuminate\Foundation\Application(getcwd());

if (is_file($configPath = $laravel->getCachedConfigPath())) {
    @unlink($configPath);
}

if (is_file($servicesPath = $laravel->getCachedServicesPath())) {
    @unlink($servicesPath);
}

if (is_file($packagesPath = $laravel->getCachedPackagesPath())) {
    @unlink($packagesPath);
}
```

**Comment by indigoram89** [comment]
@sc0Vu We need to fix this interface urgently

**Comment by legitalk** [comment]
@ttodua This issue is still present. I am unable to install CCXT 4.1.
```

> Illuminate\Foundation\ComposerScripts::postAutoloadDump
PHP Fatal error:  Declaration of ccxt\pro\Future::then($onFulfilled = null, $onRejected = null, $onProgress = null) must be compatible with React\Promise\PromiseInterface::then(?callable $onFulfilled = null, ?callable $onRejected = null): React\Promise\PromiseInterface in /home/vagrant/ohlcbot/vendor/ccxt/ccxt/php/pro/Future.php on line 15
```

**Comment by carlosmiei** [comment]
@legitalk @indigoram89 @miguilimzero can you try again now? We merged an update that should fix this issue

**Comment by indigoram89** [comment]
@carlosmiei thank you! I tried to install it in Laravel for the following way and got the error:

Screenshot 2023-11-04 at 15 10 46

The second way I tried worked fine:

Screenshot 2023-11-04 at 15 11 33

And it looks good:

Screenshot 2023-11-04 at 15 14 10

**Comment by carlosmiei** [comment]
@indigoram89 so it's working now right?

**Comment by indigoram89** [comment]
@carlosmiei yes, thank you so much!

**Comment by carlosmiei** [comment]
@indigoram89 Thanks for confirming it!

**Comment by legitalk** [comment]
@carlosmiei yes, I can also confirm it, now it is working, also important is to update composer to the latest version (2.6.5), as with older version (2.3.5) it was still not working.

**Comment by astray-heavy-arms** [comment]
Just an fyi doing this on a fresh install of laravel(10.10) with "ccxt/ccxt": "4.2.3", installed then trying to load it in my class to use but constantly getting class "ccxt" does not exist.

use ccxt\binance;
return new binance([]);

Punching in my api details and then just trying to run `php artisan optimize:clear` throws Target class [ccxt] does not exist.

Am I missing something? I though the issue I'm having sounds very similar to this and even trying @indigoram89 solution here that also doesn't work (I know outdated but worth a shot).

**Pull request merged**